### PR TITLE
vimgr OK

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

This would have been simple but for some tangled data and even more tangled headers. I elected to import the very small amount data into the voice file to save time, and to put off moving the various externs and other header stuff around until the rest of libultra's done, since it's not far off now.

### CODE REVIEW

**Review by AngheloAlf — APPROVED** [review]

File: src/libultra/io/vimgr.c [hunk 1]
```diff
@@ -1,5 +1,116 @@
...
+    u16 unk16;
+} viMesgStruct;
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/vimgr/viMgrMain.s")
+OSThread viThread;
+u8 viThreadStack[0x1000];
+OSMesgQueue viEventQueue;
+OSMesg viEventBuf[6];
+viMesgStruct viRetraceMsg;
+viMesgStruct viCounterMsg;
```
Comment (Thar0) [inline src/libultra/io/vimgr.c:18]: The correct type for these is `OSIoMesg` apparently
(https://github.com/Rozelette/libultra-mdebug/blob/master/vimgr.o.txt#L176)

File: src/libultra/io/vimgr.c [hunk 2]
```diff
@@ -1,5 +1,116 @@
...
+
+void osCreateViManager(OSPri pri) {
+    u32 prevInt;
+    OSPri newPri;
+    OSPri currentPri;
+
+    if (!__osViDevMgr.initialized) {
+        __osTimerServicesInit();
+        __additional_scanline = 0;
+        osCreateMesgQueue(&viEventQueue, viEventBuf, 5);
```
Comment (Thar0) [inline src/libultra/io/vimgr.c:32]:
```suggestion
        osCreateMesgQueue(&viEventQueue, viEventBuf, ARRAY_COUNT(viEventBuf) - 1);
```

**Review by Thar0 — CHANGES_REQUESTED** [review]

File: src/libultra/io/vimgr.c (hunk 1 above)
Comment (EllipticEllipsis) [reply src/libultra/io/vimgr.c:18]: Hm, I wonder if there's an enum for the `type`:
```c
        viRetraceMsg.hdr.type = 13;
        viRetraceMsg.hdr.pri = 0;
        viRetraceMsg.hdr.retQueue = NULL;
        viCounterMsg.hdr.type = 14;
        viCounterMsg.hdr.pri = 0;
        viCounterMsg.hdr.retQueue = NULL;
```

File: src/libultra/io/vimgr.c [hunk 3]
```diff
@@ -1,5 +1,116 @@
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/vimgr/osCreateViManager.s")
+typedef struct {
+    u16 unk00;
+    u8 unk02;
+    u32 unk04;
+    char pad[0xC];
+    u16 unk14;
+    u16 unk16;
+} viMesgStruct;
```
Comment (Thar0) [inline src/libultra/io/vimgr.c:11]: This should be removable now

File: src/libultra/io/vimgr.c [hunk 4]
```diff
@@ -1,5 +1,116 @@
...
+            osSetThreadPri(NULL, newPri);
+        }
+    }
+}
+
+void viMgrMain(void* vargs) {
+    OSMgrArgs* args;
+    static u16 viRetrace;
+    u32 addTime;
+    viMesgStruct* mesg;
```
Comment (Thar0) [inline src/libultra/io/vimgr.c:71]:
```suggestion
    OSIoMesg* mesg;
```

File: src/libultra/io/vimgr.c [hunk 5]
```diff
@@ -1,5 +1,116 @@
...
+    viRetrace = __osViGetCurrentContext()->retraceCount;
+    if (viRetrace == 0) {
+        viRetrace = 1;
+    }
+
+    args = (OSMgrArgs*)vargs;
+
+    while (true) {
+        osRecvMesg(args->eventQueue, (OSMesg)&mesg, OS_MESG_BLOCK);
+        switch (mesg->unk00) {
```
Comment (Thar0) [inline src/libultra/io/vimgr.c:84]:
```suggestion
        switch (mesg->hdr.type) {
```

File: src/libultra/io/vimgr.c [hunk 6]
```diff
@@ -1,5 +1,116 @@
...
+            case 13:
```
Comment (Thar0) [inline src/libultra/io/vimgr.c:85]:
```suggestion
            case OS_MESG_TYPE_VRETRACE:
```

File: src/libultra/io/vimgr.c [hunk 7]
```diff
@@ -1,5 +1,116 @@
...
+                    __osCurrentTime = addTime;
+                    temp = 0;
+                }
+
+                addTime = __osBaseCounter;
+                __osBaseCounter = osGetCount();
+                addTime = __osBaseCounter - addTime;
+                __osCurrentTime = __osCurrentTime + addTime;
+                break;
+            case 14:
```
Comment (Thar0) [inline src/libultra/io/vimgr.c:111]:
```suggestion
            case OS_MESG_TYPE_COUNTER:
```

**Review by Thar0 — CHANGES_REQUESTED** [review]

**Review by Thar0 — APPROVED** [review]

**Review by kyleburnette — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**